packages/vitest/src/runtime/mocker.ts

```diff
@@ -328,7 +328,29 @@ export class VitestMocker {
           const spyModule = this.spyModule
           if (!spyModule)
             throw this.createError('[vitest] `spyModule` is not defined. This is Vitest error. Please open a new issue with reproduction.')
-          const mock = spyModule.spyOn(newContainer, property).mockImplementation(() => undefined)
+
+          const primitives = this.primitives
+          const mock = spyModule.spyOn(newContainer, property).mockImplementation(function (this: any) {
+            // jest reference
+            // https://github.com/jestjs/jest/blob/2c3d2409879952157433de215ae0eee5188a4384/packages/jest-mock/src/index.ts#L678-L691
+
+            // check constructor call
+            if (this instanceof newContainer[property]) {
+              // mock each mothod of mocked class instance
+              for (const { key } of getAllMockableProperties(this, false, primitives)) {
+                const value = this[key]
+                const type = getType(value)
+                const isFunction = type.includes('Function') && typeof value === 'function'
+                if (isFunction) {
+                  // TODO: ability to restore?
```

Comment:
I checked the behavior on Jest here https://stackblitz.com/edit/jest-example-z6bfab?file=mocked-class-restore.test.js
I might be using `mockRestore` incorrectly, but Jest doesn't seem to support such use case.
I added a similar test case for Vitest https://github.com/vitest-dev/vitest/pull/4564/commits/5b60bbccea4b726ccb393f3732c0db9925678015 and `mockRestore` is currently not working either but in a different way.
Would you wish to align `mockRestore` behavior?

Reply:
I think I missed one important context https://github.com/vitest-dev/vitest/pull/3438 and the rational of this `mockRestore` override:

https://github.com/vitest-dev/vitest/blob/736b61cde36a3db008c721138111c68aaddac992/packages/vitest/src/runtime/mocker.ts#L353-L357

I don't think my current implementation considers global mock restore use case. I think I understand what's the expected behavior, so let me deal with this and I'll add an appropriate test case.

Reply:
I update the code so that constructor mocking is preserved after `vi.restoreAllMocks`, which I think is a desired behavior aligning with https://github.com/vitest-dev/vitest/pull/3438.

I made separate test cases `mocked-class-restore-explicit.test.ts` (to test `mockFn.mockRestore`) and `mocked-class-restore-all.test.ts` (to test `vi.restoreAllMocks`). The "explicit" scenario is still not aligned with Jest's behavior but I feel this use case is unusual and might be difficult to support.
I'd like to know what you think about this scenario. Thanks!